### DISCUSSION TITLE
[Question] Nearest neighbors implementation for Patchcore. FAISS vs PyTorch?

### DISCUSSION THREAD

**Discussion by thibaut-d** [open-edge-platform/anomalib#1014] [Q&A]
Hi,

I've been through the code of Patchcore. The implementation is very clean and readable. It makes it easy understand. 😊👍

I have a question regarding the implementation of the nearest neighbors algorithm. I see that you rely on PyTorch while the implementation of the author of the paper relies on FAISS. 

Is there any rational relying on one or another? Did you notice some performance issues with FAISS or is it because it is more straightforward to implement it in PyTorch ? Or another rational ?

My guess is that nearest neighbors plays a significant role in performances and power consumption for inference. 

Thanks

**Comment by samet-akcay** [comment]
Hi @thibaut-d, we wanted to write it with Torch for consistency and also deployment capabilities such as onnx openvino. It currently cannot be exported to onnx/openvino though. It needs a fix.

**Reply by peacefulotter** [reply]
Jumping up on this, have you done some benchmarks / comparisons between faiss and your implementation? Should I expect the two to be very similar in terms of performance (memory consumption, accuracy, speed) 
Thanks!

**Reply by lathashree01** [reply]
@samet-akcay Just to clarify, you mean the current Patchcore-model can not be exported to onnx? 

I am currently developing a solution with it and want to deploy it as a ONNX model in a couple of weeks. So wanted to know more on deployment capabilities wrt Patchcore.

**Comment by thibaut-d** [comment]
Thanks. I understand that.
